pkg/render/dex_config.go

```diff
@@ -0,0 +1,359 @@
+	ClientSecret() []byte
+	// ManagerURI returns the address where the Manager UI can be found. Ex: https://example.org
+	RequestedScopes() []string
+	// UsernameClaim returns the part of the JWT that represents a unique username.
+	UsernameClaim() string
+	// GroupsClaim returns the part of the JWT that represents the list of user groups.
+	GroupsClaim() string
+	baseConfig
+}
+
+type baseConfig interface {
+	// ManagerURI returns the address where the Manager UI can be found. Ex: https://example.org
+	ManagerURI() string
+	// RequiredAnnotations returns annotations that make your the pods get refreshed if any of the config/secrets change.
+	RequiredAnnotations() map[string]string
+	// RequiredSecrets returns secrets that you need to render for dex.
+	RequiredSecrets(namespace string) []*corev1.Secret
+}
+
+func NewDexRelyingPartyConfig(
+	authentication *oprv1.Authentication,
+	tlsSecret *corev1.Secret,
+	dexSecret *corev1.Secret) DexRelyingPartyConfig {
+	return &dexRelyingPartyConfig{baseCfg(authentication, tlsSecret, dexSecret, nil)}
+}
+
+func NewDexKeyValidatorConfig(
+	authentication *oprv1.Authentication,
+	tlsSecret *corev1.Secret) DexKeyValidatorConfig {
+	return &dexKeyValidatorConfig{baseCfg(authentication, tlsSecret, nil, nil)}
+}
+
+// Create a new DexConfig.
+func NewDexConfig(
+	authentication *oprv1.Authentication,
+	tlsSecret *corev1.Secret,
+	dexSecret *corev1.Secret,
+	idpSecret *corev1.Secret) DexConfig {
+	return &dexConfig{baseCfg(authentication, tlsSecret, dexSecret, idpSecret)}
+}
+
+type dexKeyValidatorConfig struct {
+	*dexBaseCfg
+}
+
+type dexConfig struct {
+	*dexBaseCfg
+}
+
+type dexRelyingPartyConfig struct {
+	*dexBaseCfg
+}
+
+// Create a struct to hold the base configuration of dex.
+func baseCfg(
+	authentication *oprv1.Authentication,
+	tlsSecret *corev1.Secret,
+	dexSecret *corev1.Secret,
+	idpSecret *corev1.Secret) *dexBaseCfg {
+
+	// If the manager domain is not a URL, prepend https://.
+	baseUrl := authentication.Spec.ManagerDomain
+	if !strings.HasPrefix(baseUrl, "http://") && !strings.HasPrefix(baseUrl, "https://") {
+		baseUrl = fmt.Sprintf("https://%s", baseUrl)
+	}
+
+	var connType connectorType
+	var issuer string
+	if authentication.Spec.OIDC != nil {
+		issuer = authentication.Spec.OIDC.IssuerURL
+		if issuer == "https://accounts.google.com" {
+			connType = connectorTypeGoogle
+		} else {
+			connType = connectorTypeOIDC
+		}
+	} else if authentication.Spec.Openshift != nil {
+		issuer = authentication.Spec.Openshift.IssuerURL
+		connType = connectorTypeOpenshift
+	}
+
+	return &dexBaseCfg{
+		authentication: authentication,
+		tlsSecret:      tlsSecret,
+		idpSecret:      idpSecret,
+		dexSecret:      dexSecret,
+		connectorType:  connType,
+		issuer:         issuer,
+		managerURI:     baseUrl,
+	}
+}
+
+type dexBaseCfg struct {
+	authentication *oprv1.Authentication
+	tlsSecret      *corev1.Secret
+	idpSecret      *corev1.Secret
+	dexSecret      *corev1.Secret
+	managerURI     string
```

Comment:
I think you should rename this to base URI and add functions like IssuerEndpoint, TokenEndpoint, and whatnot. We're encapsulating Dex here so the manager shouldn't be mentioned, and the config should be able to provide the exact endpoints that a user of this requires.